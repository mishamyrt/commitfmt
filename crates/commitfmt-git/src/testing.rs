use std::path::PathBuf;

use tempfile::TempDir;

use crate::command::run_git;
use crate::repository::Repository;
use crate::GitResult;

pub const COMMIT_HISTORY: &[&str] = &[
    "chore: initial commit",
    "feat(linter): remove description leading space rule",
    "feat(cc): trim description on parse",
    "fix: resolve parsing issue with special characters",
    "docs: update README with usage examples",
    "feat(cli): add new --verbose flag for detailed output",
    "test: add unit tests for commit message validation",
    "refactor: simplify configuration loading logic",
    "feat(format): implement auto-formatting for commit headers",
    "fix(git): handle empty repository edge case",
    "chore: update dependencies to latest versions",
    "feat: add support for custom commit templates",
    "fix(parser): correctly handle multiline commit bodies",
    "docs(api): add comprehensive API documentation",
    "feat(workspace): implement configuration inheritance",
    "test(integration): add end-to-end testing suite",
    "fix: prevent duplicate footers in formatted messages",
    "perf: optimize commit log parsing performance",
    "feat(hooks): add support for commit-msg hook",
    "ci: setup automated testing workflow",
];

pub struct TestBed {
    dir: TempDir,
    pub repo: Repository,
}

impl TestBed {
    pub const DEFAULT_BRANCH_NAME: &str = "main";
    pub const DEFAULT_USER_NAME: &str = "Test User";
    pub const DEFAULT_USER_EMAIL: &str = "test@example.com";

    /// Creates a new test bed with a fresh git repository
    pub fn empty() -> GitResult<Self> {
        let dir = TempDir::with_prefix("commitfmt-git-test-")?;
        let dir_path = dir.path();

        run_git(&["init"], dir_path)?;
        run_git(&["config", "user.name", Self::DEFAULT_USER_NAME], dir_path)?;
        run_git(&["config", "user.email", Self::DEFAULT_USER_EMAIL], dir_path)?;
        run_git(&["switch", "-c", Self::DEFAULT_BRANCH_NAME], dir_path)?;

        let repo = Repository::from_root(dir_path);

        Ok(Self { dir, repo })
    }

    /// Returns the path to the test bed
    pub fn path(&self) -> PathBuf {
        self.dir.path().to_path_buf()
    }

    /// Switches to a new branch
    pub fn switch_to_new(&self, branch_name: &str) -> GitResult<()> {
        run_git(&["switch", "-c", branch_name], self.dir.path())?;
        Ok(())
    }

    pub fn with_default_history() -> GitResult<Self> {
        let test_bed = Self::empty()?;
        for message in COMMIT_HISTORY {
            test_bed.repo.commit(message)?;
        }
        Ok(test_bed)
    }

    pub fn with_history(commits: &[&str]) -> GitResult<Self> {
        let test_bed = Self::empty()?;
        for message in commits {
            test_bed.repo.commit(message)?;
        }
        Ok(test_bed)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_new() {
        let test_bed = TestBed::empty().unwrap();
        assert!(test_bed.path().to_str().unwrap().contains("commitfmt-git-test-"));
        assert_eq!(test_bed.repo.get_branch_name(), Some("main".to_string()));
        assert_eq!(test_bed.repo.get_root(), test_bed.dir.path());
    }

    #[test]
    fn test_new_with_history() {
        let test_bed = TestBed::with_default_history().unwrap();
        assert_eq!(test_bed.repo.get_branch_name(), Some("main".to_string()));
        assert_eq!(test_bed.repo.get_root(), test_bed.dir.path());
        assert_eq!(
            test_bed.repo.run(&["rev-list", "--count", "--all"]).unwrap(),
            COMMIT_HISTORY.len().to_string()
        );
    }
}
