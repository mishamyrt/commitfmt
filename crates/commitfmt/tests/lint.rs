use commitfmt::testing::pipe_from_string;
use commitfmt_git::testing::TestBed;
use insta::assert_snapshot;
use std::process::Command;

#[test]
fn test_cli_lint_configured() {
    let test_bed = TestBed::empty().unwrap();
    let input = "
Feat  (  Test   ): Test

body text and more text

Footer: value
BREAKING CHANGES: test
"
    .trim();

    let config = r#"
[lint.header]
type-case = "lower"
description-case = "lower-first"
scope-min-length = 5
scope-case = "kebab"
min-length = 18
type-max-length = 3

[lint.body]
case = "upper-first"
full-stop = true
max-line-length = 20
max-length = 20

[lint.footer]
key-case = "lower"
min-length = 15
max-length = 20
exists = ["Issue-ID"]
"#;
    std::fs::write(test_bed.path().join("commitfmt.toml"), config).unwrap();

    let exe = env!("CARGO_BIN_EXE_commitfmt");

    let mut cmd = Command::new(exe);
    cmd.arg("--lint");
    cmd.stdin(pipe_from_string(input));
    cmd.current_dir(test_bed.path());

    let output = cmd.output().unwrap();
    assert!(!output.status.success());

    let result = String::from_utf8(output.stdout).unwrap();

    assert_snapshot!("cli_configured", result);
}
