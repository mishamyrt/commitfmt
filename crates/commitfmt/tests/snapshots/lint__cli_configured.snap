---
source: crates/commitfmt/tests/lint.rs
assertion_line: 54
expression: result
---
- Type case is inconsistent. Expected: lower [type-case]
- Type is longer than 3 characters [type-max-length]
- Scope case is inconsistent. Expected: kebab [scope-case]
- Scope is shorter than 5 characters [scope-min-length]
- Description case is inconsistent. Expected: lower-first [description-case]
- Header is shorter than 18 characters [min-length]
- Body line is longer than 20 characters [max-line-length]
- Body is longer than 20 characters [max-length]
- Body is not ended with a full stop [full-stop]
- Body case is inconsistent. Expected: upper-first [case]
- Footer 'Issue-ID' is required but not found [exists]
- Footer 'BREAKING CHANGES' length is longer than 20 characters [max-length]
- Message contains breaking changes footer but no exclamation mark [breaking-exclamation]
- Footer key case is inconsistent. Expected: lower [key-case]
- Footer 'Footer' length is less than 15 characters [min-length]

Found 15 problems
